sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/manager/DownloadManagerReleaseDownloader.java

```diff
@@ -0,0 +1,228 @@
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOAD_ID;
+
+public class DownloadManagerReleaseDownloader implements ReleaseDownloader {
+
+    private final Context mContext;
+
+    private final ReleaseDetails mReleaseDetails;
+
+    private final ReleaseDownloader.Listener mListener;
+
+    private long mDownloadId = INVALID_DOWNLOAD_IDENTIFIER;
+
+    /**
+     * Current task to check download state and act on it.
+     */
+    private DownloadManagerUpdateTask mUpdateTask;
+
+    /**
+     * Current task inspecting the latest release details that we fetched from server.
+     */
+    private DownloadManagerRequestTask mRequestTask;
+
+    public DownloadManagerReleaseDownloader(@NonNull Context context, @NonNull ReleaseDetails releaseDetails, @NonNull ReleaseDownloader.Listener listener) {
+        mContext = context;
+        mReleaseDetails = releaseDetails;
+        mListener = listener;
+    }
+
+    private DownloadManager getDownloadManager() {
+        return (DownloadManager) mContext.getSystemService(DOWNLOAD_SERVICE);
+    }
+
+    private long getDownloadId() {
+        if (mDownloadId == INVALID_DOWNLOAD_IDENTIFIER) {
+            mDownloadId = SharedPreferencesManager.getLong(PREFERENCE_KEY_DOWNLOAD_ID, INVALID_DOWNLOAD_IDENTIFIER);
+        }
+        return mDownloadId;
+    }
+
+    public ReleaseDetails getReleaseDetails() {
+        return mReleaseDetails;
+    }
+
+    @Override
+    public void resume() {
+
+        /* If there is active downloads. */
+        long downloadId = getDownloadId();
+        if (downloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+            update();
+        } else {
+            request();
+        }
+    }
+
+    @Override
+    public void delete() {
+        if (mRequestTask != null) {
+            mRequestTask.cancel(true);
+            mRequestTask = null;
+        }
+        if (mUpdateTask != null) {
+            mUpdateTask.cancel(true);
+            mUpdateTask = null;
+        }
+        long downloadId = getDownloadId();
+        if (downloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+            remove(downloadId);
+        }
+        SharedPreferencesManager.remove(PREFERENCE_KEY_DOWNLOAD_ID);
+    }
+
+    /**
+     * Start new download.
+     */
+    private void request() {
+        mRequestTask = AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerRequestTask(this));
+    }
+
+    /**
+     * Update the state on current download.
+     */
+    private void update() {
+        mUpdateTask = AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerUpdateTask(this));
+    }
+
+    private void remove(long downloadId) {
+        AppCenterLog.debug(LOG_TAG, "Removing download and notification id=" + downloadId);
+        AsyncTaskUtils.execute(LOG_TAG, new DownloadManagerRemoveTask(mContext, downloadId));
+    }
+
+    @WorkerThread
+    void onRequest(DownloadManagerRequestTask task) {
```

Comment:
General comment, most of the code that was moved out of Distribute was using synchronized before checking state changes.

How is the threading story now? Please review possible race conditions (synchronized in different classes will not work I guess, before there was only 1 lock and it was the distribute class, how does that work now?).